perfkitbenchmarker/providers/openstack/os_disk.py

```diff
@@ -24,96 +24,198 @@
-                is_deleted = volume is None
-                time.sleep(sleep)
-                sleep_count += 1
-                if sleep_count == 10:
-                    sleep = 5
-        except NotFound:
-            logging.info('Volume %s not found, might have been already deleted'
-                         % self._disk.id)
-
-    def _Exists(self):
-        from novaclient.exceptions import NotFound
-        try:
-            volume = self.__nclient.volumes.get(self._disk.id)
-            return volume and volume.status in ('available', 'in-use',
-                                                'deleting',)
-        except NotFound:
-            return False
-
-    def Attach(self, vm):
-        self.attached_vm_name = vm.name
-        self.attached_vm_id = vm.id
-
-        result = self.__nclient.volumes.create_server_volume(vm.id,
-                                                             self._disk.id,)
-        self.attach_id = result.id
-
-        volume = None
-        is_unattached = True
-        while is_unattached:
-            time.sleep(1)
-            volume = self.__nclient.volumes.get(result.id)
-            if volume:
-                is_unattached = not(volume.status == "in-use"
-                                    and volume.attachments)
-
-        for attachment in volume.attachments:
-            if self.attach_id == attachment.get('volume_id'):
-                self.device = attachment.get('device')
-                return
-
-        raise errors.Error("Couldn't not attach volume to %s" % vm.name)
-
-    def GetDevicePath(self):
-        return self.device
-
-    def Detach(self):
-        self.__nclient.volumes.delete_server_volume(self.attached_vm_id,
-                                                    self.attach_id)
+
+class OpenStackDisk(disk.BaseDisk):
+  """
+  Base class for OpenStack disks.
+  """
+  def __init__(self, disk_spec, name):
+    super(OpenStackDisk, self).__init__(disk_spec)
+    self.name = name
+
+  def _Create(self):
+    pass
+
+  def _Delete(self):
+    pass
+
+  def Attach(self, vm):
+    pass
+
+  def Detach(self):
+    pass
+
+
+class OpenStackBlockStorageDisk(OpenStackDisk):
+  """
+  OpenStackBlockStorageDisk is an OpenStackDisk object to represent a
+  remote block storage volume that can be attached to an instance.
+  """
+
+  def __init__(self, disk_spec, name, zone):
+    super(OpenStackBlockStorageDisk, self).__init__(disk_spec, name)
+    self.__nclient = os_utils.NovaClient()
+    self.attached_vm_name = None
+    self.attached_vm_id = -1
+    self.name = 'cinder-volume-%s' % name
+    self.zone = zone
+    self.device = ""
+    self._disk = None
+
+  def _Create(self):
+    self._disk = self.__nclient.volumes.create(self.disk_size,
+                                               display_name=self.name,
+                                               availability_zone=self.zone,)
+    is_unavailable = True
+    while is_unavailable:
```

Comment:
You should use `@Retry` from vm_util.py to handle the retries. That will automatically back off the timing, and raise an exception if it tries too many times.